trie/triestate/state.go

```diff
@@ -16,13 +16,225 @@
 // The value refers to the original content of state before the transition
 // is made. Nil means that the state was not present previously.
 type Set struct {
 	Accounts   map[common.Address][]byte                 // Mutated account set, nil means the account was not present
 	Storages   map[common.Address]map[common.Hash][]byte // Mutated storage set, nil means the slot was not present
-	Incomplete map[common.Address]struct{}               // Indicator whether the storage slot is incomplete due to large deletion
+	Incomplete map[common.Address]struct{}               // Indicator whether the storage is incomplete due to large deletion
+	size       common.StorageSize                        // Approximate size of set
+}
+
+// New constructs the state set with provided data.
+func New(accounts map[common.Address][]byte, storages map[common.Address]map[common.Hash][]byte, incomplete map[common.Address]struct{}) *Set {
+	return &Set{
+		Accounts:   accounts,
+		Storages:   storages,
+		Incomplete: incomplete,
+	}
+}
+
+// Size returns the approximate memory size occupied by the set.
+func (s *Set) Size() common.StorageSize {
+	if s.size != 0 {
+		return s.size
+	}
+	for _, account := range s.Accounts {
+		s.size += common.StorageSize(common.HashLength + len(account))
+	}
+	for _, slots := range s.Storages {
+		for _, val := range slots {
+			s.size += common.StorageSize(common.HashLength + len(val))
+		}
+		s.size += common.StorageSize(common.HashLength)
+	}
+	s.size += common.StorageSize(common.HashLength * len(s.Incomplete))
+	return s.size
+}
+
+// context wraps all fields for executing state diffs.
+type context struct {
+	prevRoot    common.Hash
+	postRoot    common.Hash
+	accounts    map[common.Address][]byte
+	storages    map[common.Address]map[common.Hash][]byte
+	accountTrie Trie
+	nodes       *trienode.MergedNodeSet
+}
+
+// Apply traverses the provided state diffs, apply them in the associated
+// post-state and return the generated dirty trie nodes. The state can be
+// loaded via the provided trie loader.
+func Apply(prevRoot common.Hash, postRoot common.Hash, accounts map[common.Address][]byte, storages map[common.Address]map[common.Hash][]byte, loader TrieLoader) (map[common.Hash]map[string]*trienode.Node, error) {
+	tr, err := loader.OpenTrie(postRoot)
+	if err != nil {
+		return nil, err
+	}
+	ctx := &context{
+		prevRoot:    prevRoot,
+		postRoot:    postRoot,
+		accounts:    accounts,
+		storages:    storages,
+		accountTrie: tr,
+		nodes:       trienode.NewMergedNodeSet(),
+	}
+	for addr, account := range accounts {
+		var err error
+		if len(account) == 0 {
+			err = deleteAccount(ctx, loader, addr)
+		} else {
+			err = updateAccount(ctx, loader, addr)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to revert state, err: %w", err)
+		}
+	}
+	root, result := tr.Commit(false)
+	if root != prevRoot {
+		return nil, fmt.Errorf("failed to revert state, want %#x, got %#x", prevRoot, root)
```

Comment:
I guess this clause will implicitly catch the issue if there are incomplete accounts?

Reply:
Exactly, not only incomplete accounts, all invalid state history should be captured by this check.

The trie(account + storage tries) is supposed to be reverted to previous state, otherwise, bail out.